### PR TITLE
Make StateVariable an IR class that accepts IR::Member. Use ICastable for some nodes.

### PR DESCRIPTION
**Pull request by fruffy**
Some preparation for #3519. Remove implicit casting from StateVariable and explicitly require IR::Member as initializer argument. 

This required a small change to [ir-generator.ypp](https://github.com/p4lang/p4c/pull/3741/files#diff-0ee13e28f9c23c0a103c9e59759d5e112d41813c3dc330b66ac922c596c77177). Those operators were not part of the syntax.

### CODE REVIEW

File: backends/p4tools/modules/testgen/lib/test_backend.cpp [hunk 1]
```diff
@@ -64,8 +63,9 @@ const Model *TestBackEnd::computeConcolicVariables(const ExecutionState *executi
             const auto *concolicAssignment = resolvedConcolicVariable.second;
             const IR::Expression *pathConstraint = nullptr;
             // We need to differentiate between state variables and expressions here.
-            if (std::holds_alternative<const StateVariable>(concolicVariable)) {
-                pathConstraint = new IR::Equ(std::get<const StateVariable>(concolicVariable),
+            // We need to differentiate between state variables and expressions here.
```
Comment (pkotikal) [inline backends/p4tools/modules/testgen/lib/test_backend.cpp:66]: Repeated comment.

File: backends/p4tools/modules/testgen/test/z3-solver/expressions.cpp [hunk 2]
```diff
@@ -125,9 +125,9 @@ void test(const IR::Expression *expression, const IR::AssignmentStatement *varia
     Model model = *solver.getModel();
     ASSERT_EQ(model.size(), 2u);
 
-    ASSERT_EQ(model.count(variableValue->left), 1u);
+    ASSERT_EQ(model.count(variableValue->left->checkedTo<IR::Member>()), 1u);
```
Comment (pkotikal) [inline backends/p4tools/modules/testgen/test/z3-solver/expressions.cpp:128]: `1U`

**Review by pkotikal — APPROVED** [review]
LGTM.

File: backends/p4tools/modules/testgen/lib/concolic.h [hunk 3]
```diff
@@ -28,7 +27,7 @@ namespace P4Tools::P4Testgen {
 /// are present, but not expressions. The reason expressions need to be keys is that sometimes
 /// entire expressions are mapped to a particular constant.
 using ConcolicVariableMap =
-    ordered_map<std::variant<const StateVariable, const IR::Expression *>, const IR::Expression *>;
+    std::map<std::variant<const IR::StateVariable, const IR::Expression *>, const IR::Expression *>;
```
[the remark this answers is not in the document]
Comment (fruffy) [reply backends/p4tools/modules/testgen/lib/concolic.h:30]: Nope, that was a rebase mistake, good catch!

**Comment by fruffy** [comment]
@mbudiu-vmw This adds new operators to the syntax in the ir-generator yacc file. Any potential issues that could arise here?

**Comment by mbudiu-vmw** [comment]
> @mbudiu-vmw This adds new operators to the syntax in the ir-generator yacc file. Any potential issues that could arise here?

You probably mean that you have a new .def file. That should always work.

**Comment by fruffy** [comment]
> > @mbudiu-vmw This adds new operators to the syntax in the ir-generator yacc file. Any potential issues that could arise here?
> 
> You probably mean that you have a new .def file. That should always work.

I was referring to the edits here:
https://github.com/p4lang/p4c/pull/3741/files#diff-0ee13e28f9c23c0a103c9e59759d5e112d41813c3dc330b66ac922c596c77177R288

The syntax was missing `->` and `*` which need to be overridden for this particular IR construct.

**Comment by mbudiu-vmw** [comment]
> > > @mbudiu-vmw This adds new operators to the syntax in the ir-generator yacc file. Any potential issues that could arise here?
> > 
> > 
> > You probably mean that you have a new .def file. That should always work.
> 
> I was referring to the edits here: https://github.com/p4lang/p4c/pull/3741/files#diff-0ee13e28f9c23c0a103c9e59759d5e112d41813c3dc330b66ac922c596c77177R288
> 

The file in question has been written entirely by @ChrisDodd

**Pull request merged**